packages/devcmd/src/process/ProcessExecutor.ts

```diff
@@ -39,161 +53,231 @@ class SafeConsoleLike implements ConsoleLike {
+   *
+   * The "nonZeroExitCodeHandling" option of the given {@link ProcessInfo}s determines
+   * whether this method throws on a non-zero exit code. This option applies to each process
+   * _individually_, so in a single call some processes can ignore non-zero exit codes while
+   * others throw an error.
+   *
+   * **Example:** Printing node and npm version to the console:
+   * ```
+   * await execPipedParallel({
+   *   nodeVersion: { command: "node", args: ["-v"] },
+   *   npmVersion: { command: "npm", args: ["--version"] },
+   * });
+   * ```
    *
-   * @param processMap A map linking process ids to {@link ProcessInfo} instances
-   * @returns A promise that resolves on success and rejects on error
+   * @param processMap A map of a caller-defined process ID to {@link ProcessInfo} for each
+   * process to execute. The given process IDs are used for console output and in the
+   * returned map of exit status infos.
    *
+   * @returns A promise resolving to a map of the process ID (from `processMap`) to exit
+   * status info, once all processes have exited
+   */
+  async execPipedParallel<T extends { [id: string]: ProcessInfo }>(
+    processMap: T
+  ): Promise<{ [id in keyof T]: NodeExitInfo }>;
+
+  /**
+   * Executes multiple processes in parallel and pipes their outputs (stdout/stderr) to the
+   * log/error level of this instance's {@link ConsoleLike}.
    *
-   * @example
-   * <caption>Printing node and npm version to the console</caption>
+   * The "nonZeroExitCodeHandling" option of the given {@link ProcessInfo}s determines
+   * whether this method throws on a non-zero exit code. This option applies to each process
+   * _individually_, so in a single call some processes can ignore non-zero exit codes while
+   * others throw an error.
+   *
+   * **Example:** Printing node and npm version to the console:
    * ```
    * await execPipedParallel({
-   *   nodeVersion: {
-   *     command: "node",
-   *     args: ["-v"],
-   *   },
-   *   npmVersion: {
-   *     command: "npm",
-   *     args: ["--version"],
-   *   },
+   *   nodeVersion: { command: "node", args: ["-v"] },
+   *   npmVersion: { command: "npm", args: ["--version"] },
    * });
    * ```
+   *
+   * @param processList A list of one {@link ProcessInfo} for each process to execute.
+   * The indexes in this array are used for console output.
+   *
+   * @returns A promise resolving to a list of exit status infos, one for each process in
+   * the same order as `processList`, once all processes have exited
    */
-  async execPipedParallel(processMap: { [id: string]: ProcessInfo } | { [id: number]: ProcessInfo }): Promise<void> {
-    const processEntries = Object.entries(processMap);
-    this.printNotice(`Beginning parallel execution of ${processEntries.length} processes...`);
+  async execPipedParallel(processList: ProcessInfo[]): Promise<NodeExitInfo[]>;
+
+  /** Implementation for the two overload signatures above. */
```

Comment:
What is the benefit of this? 😅 maybe I am missing something, kind of hard to read in the GitHub diff but I cant checkout the code right now ^-^

Reply:
Since we now return one result per process from `execPipedParallel`, it's important to make it easy (from a caller's point of view) to match the results to the processes passed in.

So if you pass a map (=object) of ProcessInfo, you get a map (=object) of NodeExitInfo, which uses the same keys that your input map did.
If you pass an array of ProcessInfo, you get an array of NodeExitInfo whose order corresponds to that if your input array.

I first tried specifying this behavior in the method signature using [conditional types](https://www.typescriptlang.org/docs/handbook/2/conditional-types.html), but I couldn't TS to accept the implementation as satisfying the conditional return type.
So I switched over to using overloads instead. In this case, consumers of the code only see the two overload signatures on lines 146 and 147; the implementation signature on line 150 is hidden automatically (I was positively surprised by this latter point).

----

For reference, the conditional return type signature that I tried looked something like:

```ts
async execPipedParallel<T extends { [id: string]: ProcessInfo } | ProcessInfo[]>
  (processMapOrList: T): 
Promise<T extends ProcessInfo[] ? NodeExitInfo[] : { [id in keyof T]: NodeExitInfo }> {
  // . . .
}
```

As far as I could figure out, basically no implementation can satisfy this because I couldn't make the compiler narrow `T` to `ProcessInfo[]`. All I could do is narrow the type of `processMapOrList` with runtime checks (e.g. `Array.isArray(processMapOrList)`), but that's not enough.
In the end, I gave up and chose the overload signature approach instead.